controllers/etcdcluster_controller.go

```diff
@@ -83,14 +86,130 @@ func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 			log.Error(err, "unable to create Service", "namespace", service.Namespace, "cluster", cluster.Name)
 			return ctrl.Result{}, err
 		}
+		log.V(1).Info("Created Service", "service name", service.Name)
 		return ctrl.Result{}, nil
 	}
+	log.V(2).Info("Service exists", "service name", service.Name)
+
+	peers := &etcdv1alpha1.EtcdPeerList{}
+	if err := r.List(ctx, peers, client.MatchingFields{"spec.clusterName": cluster.Name}); err != nil {
+		log.Error(err, "unable to list peers", "namespace", service.Namespace, "cluster", cluster.Name)
+		return ctrl.Result{}, err
+	}
+
+	// Assume we can't communicate with the cluster
+	log.V(1).Info("Unable to communicate with etcd cluster")
+	if int32(len(peers.Items)) < *cluster.Spec.Replicas {
+
+		// Create more peers
+		peerName := nextAvailablePeerName(cluster, peers.Items)
```

Comment:
does kubebuilder (or the controller-runtime lib?) guarantee that we are executing at most one call to `Reconcile()` at once? things may go very wrong if that's not the case

Reply:
@munnerz? My expectation when writing this code is very much that the controller runtime will mean we cannot have more than one copy running at once.

Reply:
The manager process has a leader election mechanism which ensures that there's only one of them running in the cluster:
```
.//bin/manager --help
Usage of ./bin/manager:
  -enable-leader-election
    	Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.
```

And I think there *may* be multiple workers / goroutines running Reconcile(), but they are all popping items from a work queue which de-duplicates the `namespace/name` keys.

So there should only ever be one Reconcile() dealing with any particular EtcdCluster or EtcdPeer.

This allows the manager to be reconciling multiple clusters in parallel.

Reply:
from a quick skim of the code, i can't see anything that obviously deduplicates keys in the work queue, but i haven't looked in depth. however it looks like we're using the default concurrent reconcile value, which is 1, so there shouldn't be any concurrency issues here